pipeline/pipeline/task.py

```diff
@@ -69,11 +90,17 @@ def handler(match_obj):
                     raise KeyError(f"undefined resource '{r_uid}' in command '{command}'.\n"
                                    f"Hint: resources must be from the same pipeline as the current task.")
                 if r._source != self:
-                    self._inputs.add(r)
+                    self._add_inputs(r)
                     if r._source is not None:
+                        if r not in r._source._mentioned:
+                            name = r._source._resources_inverse
+                            raise Exception(f"undefined resource '{name}'\n"
+                                            f"Hint: resources must be defined within "
+                                            "the task methods 'command' or 'declare_resource_group'")
                         self._dependencies.add(r._source)
+                        r._source._add_outputs(r)
                 else:
-                    self._outputs.add(r)
+                    self._mentioned.add(r)
```

Comment:
This is related to the implicit attribute creation, but I kind of feel nervous about implicitly adding resources. We have 2 modes: `if t2: don't add` if `t1: add` for `t1 = Task(pipeline=...)`,` t2 = Task(pipeline=...)`, and `t1.command` vs `t2.command`.